### ISSUE TITLE
Static Assets Issue: When Deploying a Next.js Container

### ISSUE THREAD

**Issue by hanbao-dev** [anomalyco/sst#5265]
Not entirely sure what's going on here, but every time I deploy and the new instance is spinning up, this issue appears briefly. It will either throw a client-side exception or cause the assets to be missing. I believe the optimal solution might be to have a partial deployment that sends assets to CloudFront first, rather than bundling them with the container. I'm not sure how best to achieve this with SST.

<img width="1291" alt="Screenshot 2024-12-28 at 11 53 15 PM" src="https://github.com/user-attachments/assets/b335c870-891d-4e5c-a27b-b323c46206fc" />

<img width="992" alt="Screenshot 2024-12-28 at 11 53 34 PM" src="https://github.com/user-attachments/assets/80cba660-066c-4aa9-9bba-beff7d147de6" />

**Comment by jayair** [comment]
Can you share your sst config for this?

**Comment by hanbao-dev** [comment]
> Can you share your sst config for this?

```
/// <reference path="./.sst/platform/config.d.ts" />

export default $config({
  app(input) {
    return {
      name: "pigs-go-marching",
      removal: input?.stage === "production" ? "retain" : "remove",
      protect: ["production"].includes(input?.stage),
      home: "aws",
    }
  },
  async run() {
    await import("./infra/pgm-vpc")
    await import("./infra/pgm-cluster")
    await import("./infra/pgm-nextjs")
    await import("./infra/pgm-secrets")
    await import("./infra/pgm-redis")
    await import("./infra/pgm-rds")
    await import("./infra/pgm-dev")
  },
})
```

//pgm-vpc

```
import { vpc } from "./pgm-vpc"

export const cluster = new sst.aws.Cluster("PGMCluster", {
  vpc,
  forceUpgrade: "v2",
})
```

//pgm-dev
```
import { rds } from "./pgm-rds"

export const studio = new sst.x.DevCommand("Studio", {
  link: [rds],
  dev: {
    directory: "packages/pgm-core",
    command: "bunx drizzle-kit studio",
    autostart: true,
  },
})
```
//pgm-nextjs
```
import { blastProjectId, nonceJWTSecret } from "./pgm-secrets"

import { cluster } from "./pgm-cluster"
import { rds } from "./pgm-rds"
import { redis } from "./pgm-redis"

cluster.addService("PgmNextJsService", {
  image: {
    dockerfile: "apps/pgm-nextjs/Dockerfile",
  },
  public: {
    domain: "mydomain",
    ports: [{ listen: "443/https", forward: "3000/http" }],
  },
  dev: {
    directory: "apps/pgm-nextjs",
    command: "bun dev",
  },
  cpu: "1 vCPU",
  memory: "4 GB",
  link: [nonceJWTSecret, blastProjectId, redis, rds],
})
```
//pgm-rds
```
import { vpc } from "./pgm-vpc"

export const rds =
  $app.stage === "development"
    ? new sst.aws.Postgres("PGMRds", { vpc })
    : sst.aws.Postgres.get("PGMRds", {
        id: "pigs-go-marching-development-pgmrdsinstance",
      })

```
//pgm-redis
```
import { vpc } from "./pgm-vpc"

export const redis =
  $app.stage === "development"
    ? new sst.aws.Redis("PGMRedis", { vpc })
    : sst.aws.Redis.get("PGMRedis", "pigs-go-marching-development-pgmredis")
```

//pgm-vpc
```
export const vpc =
  $app.stage === "development"
    ? new sst.aws.Vpc("PGMVpc", {
        bastion: true,
      })
    : sst.aws.Vpc.get("PGMVpc", "vpc-022ce19f133595150")
```

//pgm-secrets
```
export const nonceJWTSecret = new sst.Secret("NonceJWTSecret")
export const blastProjectId = new sst.Secret("BlastProjectId")
```

**Comment by hanbao-dev** [comment]
DockerFile

FROM node:20.11.1-slim AS base

RUN apt-get update && apt-get install -y curl unzip \
    && curl -fsSL https://bun.sh/install | BUN_INSTALL=/usr/local bash \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

FROM base AS deps
WORKDIR /app

COPY package.json bun.lockb ./
COPY apps/pgm-nextjs/package.json ./apps/pgm-nextjs/
COPY packages/pgm-core/package.json ./packages/pgm-core/
COPY apps/pgm-nextjs/sst-env.d.ts* ./apps/pgm-nextjs/

RUN bun install

FROM base AS builder
WORKDIR /app

ARG SST_RESOURCE_NonceJWTSecret
ARG SST_RESOURCE_BlastProjectId
ARG SST_RESOURCE_PGMRds
ARG SST_RESOURCE_PGMRedis

COPY --from=deps /app/node_modules ./node_modules
COPY . .

ENV NEXT_TELEMETRY_DISABLED=1

WORKDIR /app/apps/pgm-nextjs
RUN bun run build

FROM base AS runner
WORKDIR /app

ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

RUN mkdir -p ./apps/pgm-nextjs/public \
    ./apps/pgm-nextjs/.next \
    ./apps/pgm-nextjs/.next/cache

RUN chown -R nextjs:nodejs ./apps/pgm-nextjs

COPY --from=builder /app/apps/pgm-nextjs/public ./apps/pgm-nextjs/public

COPY --from=builder /app/apps/pgm-nextjs/.next/standalone ./
COPY --from=builder /app/apps/pgm-nextjs/.next/static ./apps/pgm-nextjs/.next/static

RUN chown -R nextjs:nodejs ./apps/pgm-nextjs/.next

USER nextjs

EXPOSE 3000

ENV NO_COLOR=1
ENV FORCE_COLOR=0
ENV PORT=3000
ENV HOSTNAME="0.0.0.0"

CMD ["bun", "run", "apps/pgm-nextjs/server.js"]

**Comment by hanbao-dev** [comment]
@jayair hard to test since I can't always reproduce, but I put my container behind cloudfront using the new vpc origins. Seems like it may have fixed it due to caching assets. Will continue to update as I deploy more to see if it happens again. 

Cloudfront vpc origins should be added to the router.

**Comment by hanbao-dev** [comment]
Update: Confirmed there are still some chunks failing to fetch when using cloudfront vpc origins

**Comment by jayair** [comment]
I mean isn't this sorta expected? The assets are in the container when you deploy it as a container right?

**Comment by hanbao-dev** [comment]
> I mean isn't this sorta expected? The assets are in the container when you deploy it as a container right?

Is there a way to deploy the assets separately to cloudfront? Do you think doing that would help the issue?

**Comment by jayair** [comment]
Yeah I mean you could. The Cluster component doesn't do anything for that specifically. You'll probably need to handle it on your own.

Closing this for now.
